src/Interpreters/Cache/FileCache.cpp

```diff
@@ -418,6 +451,11 @@ FileSegmentsHolder FileCache::getOrSet(const Key & key, size_t offset, size_t si
     {
         fillHolesWithEmptyFileSegments(file_segments, key, range, /* fill_with_detached */false, settings, cache_lock);
     }
+
+    auto & on_key_eviction_func = files[key].on_key_eviction_func;
```

Comment:
Put this into `CreateFileSegmentSettings` instead and execute in `addCell` when a new key is created.